bin/explain.bash

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env bash
+
+# This script takes care of explaining Compliant Kubernetes Apps config and secrets.
+# It's not to be executed on its own but rather via 'ck8s explain'.
+
+set -euo pipefail
+
+declare here
+here="$(dirname "$(readlink -f "$0")")"
+declare root
+root="$(dirname "${here}")"
+
+# shellcheck source=bin/common.bash
+source "${here}/common.bash"
+
+usage() {
+  log_error "error: invalid argument: \"${1:-}\""
+  log_error "usage: ck8s explain <config|secrets> [key.to.parameter]"
+  exit 1
+}
+
+dereference() {
+  # shellcheck disable=SC2016
+  yq4 '. as $root | with(
+    .. | select(key == "$ref");
+    . = "$root." + (
+      sub("#/", "") | split "/" | with(.[]; . = "\"" + . + "\"" ) | join "."
+    )
+  ) | with(
+    .. | select(has "$ref");
+    . = eval(."$ref")
+  )' "$1"
+}
+
+hasreference() {
+  # shellcheck disable=SC2016
+  yq4 '[.. | select(key == "$ref")] | length != 0' "$1"
+}
+
+recvdereference() {
+  local data
+  data="$(dereference "$1")"
+
+  while [[ "$(hasreference - <<< "${data}")" == true ]]; do
+    data="$(dereference - <<< "${data}")"
+  done
+
+  echo "${data}"
+}
+
+explain() {
+  local data
+  data="$(recvdereference "$1")"
+
+  if [[ -n "${2:-}" ]]; then
+    local -a path
+    readarray -t path <<< "$(yq4 'split "." | .[]' <<< "$2")"
+
+    for key in "${path[@]}"; do
+      case "$(yq4 ".type" <<< "${data}")" in
+      array)
+        data="$(yq4 ".items.properties.${key}" <<< "${data}")" ;;
+      object)
+        data="$(yq4 ".properties.${key}" <<< "${data}")" ;;
+      *)
+        echo "unable to navigate to ${key} on path to $2 found unnavigable type: $(yq4 ".type" <<< "${data}")" >&2
+        exit
+      esac
+
+      if [[ "${data}" == null ]]; then
+        echo "unable to navigate to ${key} on path to $2 found nothing" >&2
+        exit
+      fi
+    done
+  fi
+
+  local title desc type
+
+  title="$(yq4 ".title // \"${2:-root}\"" <<< "${data}")"
+  desc="$(yq4 '.description // "no description"' <<< "${data}")"
+  type="$(yq4 '{"type": .type}' <<< "${data}")"
+
+  echo -e "\e[1m${title}\e[22m"
+  echo
+  echo "$desc"
+  echo
+  yq4 '{"type": .type}' <<< "${data}"
+  echo
+  case "${type##type: }" in
+  array)
+    yq4 '{"properties": .items.properties | keys}' <<< "${data}"
+    echo
+    ;;
+  object)
+    yq4 '{"properties": .properties | keys}' <<< "${data}"
+    echo
+    ;;
+  esac
+  yq4 '{"defaults": .default}' <<< "${data}"
```

Comment:
Worth suppressing `defaults: null`, or leave that as motivation for adding that to the schema?

I'm less sure everything will have defaults, especially higher level maps where the defaults may be contained in each property.

Reply:
Yes, for types `object` and `array` without properties/items.